### ISSUE TITLE
Model not training properly when using ImageRecordIter

### ISSUE THREAD

**Issue by CarlosUziel** [dmlc/mxnet#4245]
Hi,

I'm using R and MXNET to predict 7 classes from a certain dataset. I decided to change the way my model receives the training data from directly loading a data.frame from the original dataset file to creating the corresponding ImageRecordIter. However, when the model is trained using this iterator, instead of the dataset array, I always get the same predicted label when using predict().

The dataset is made of 35887 images of 48x48x1 each.

The following code exemplifies the creation of the .lst and .rec for the iterator. Images are already written on disk. Also, the naming for the validation set (instead of test set) is just for convenience.

```R
data <- read.csv("Kaggle_facial_expressions.csv")

# create a file record containing index, label, image path

# image files list
image_files <- mixedsort(list.files("./images", pattern="*.jpeg", full.names=FALSE))

# labels of each image
labels <- data[,1]

# num
index <- array(1:length(image_files))

# combines index, labels and image_files
image_list <- data.frame(index, labels, image_files)

# Shuffled data.frame
shuffled <- image_list[sample(1:nrow(image_list)),]

# separate between train and validation sets
train <- round(nrow(shuffled) * 0.9)
val <- round(nrow(shuffled) * 0.1)

write.table(  shuffled[1:train,], 
              file = "train.lst", 
              sep = "\t", 
              row.names = FALSE, 
              col.names = FALSE,
              quote = FALSE)

write.table(  shuffled[(train+1):(train+val),],
              file = "val.lst",
              sep = "\t", 
              row.names = FALSE, 
              col.names = FALSE,
              quote = FALSE)

# after that, run the following
# For more details, run ./bin/im2rec.
system('~/mxnet/bin/im2rec train.lst ./images/ train.rec color=0 resize=48')
system('~/mxnet/bin/im2rec val.lst ./images/ val.rec color=0 resize=48')
```

In the following extracted code you can see how I'm using the iterator.

```R
get_iterator <- function(data_dir = "./", batch_size = 100, data.shape) {
  data.shape <- data.shape
  train <- mx.io.ImageRecordIter(
    path.imgrec             = paste0(data_dir, "train.rec"),
    path.imglist            = paste0(data_dir, "train.lst"),
    batch.size              = batch_size,
    data.shape              = data.shape,
    mean.img                = "xxxxx"
  )
  
  val <- mx.io.ImageRecordIter(
    path.imgrec     = paste0(data_dir, "val.rec"),
    path.imglist    = paste0(data_dir, "val.lst"),
    batch.size              = batch_size,
    data.shape              = data.shape,
    mean.img                = "xxxxx"
  )
  ret <- list(train = train, value = val)
}

# get iterator
iter <- get_iterator(data_dir, 50, c(48, 48, 1))
train <- iter$train
val <- iter$val

model <- mx.model.FeedForward.create(
  X                  = train,
  eval.data          = val,
  ctx                = devices,
  symbol             = net,
  eval.metric        = mx.metric.accuracy,
  num.round          = 10,
  learning.rate      = 0.01,
  momentum           = 0.9,
  wd                 = 0.00001,
  array.batch.size   = 32,
  batch.end.callback = mx.callback.log.train.metric(50),
  initializer        = mx.init.Xavier(factor_type = "in", magnitude = 2.34),
  optimizer          = "sgd"
)

# Predict labels
predicted <- predict(model, val, mx.gpu())
# Assign labels
predicted_labels <- max.col(t(predicted)) - 1
```

After training the model with these iterators, I have run predict() with either the val array and the val iterator, obtaining the same result, so my intuition is that something is going on when the training is performed using these iterators. Also, the training accuracy is much lower when the training is done with iterators (shouldn't it be more or less the same?), and it even gets stuck at the same accuracy, which may be a consequence of always predicting the same label afterwards.

I would appreciate any insights or any help you could provide to further debug this behaviour.

Thank you for reading.

**Comment by CarlosUziel** [comment]
Sorry for not updating this issue. Both training and predicting using the ImageRecord iterators solved my problem.
